File: Python/10.py
# https://adventofcode.com/2022/day/10

with open("../data/10.txt") as file:
    data = file.read().splitlines()

signals = []
current_cycle = 1  # 1 = during first cycle, 2 = during second cycle etc.
x = 1
for instruction in data:
    if instruction == "noop":
        signals.append(x*current_cycle)
        current_cycle += 1

    elif instruction.startswith("addx"):
        signals.append(x*current_cycle)
        current_cycle += 1
        signals.append(x*current_cycle)
        current_cycle += 1

        amount = int(instruction[5:])
        x += amount

print("Part 1")
print(signals[19] + signals[59] + signals[99] + signals[139] + signals[179] + signals[219])

print("Part 2")

def what_to_print(x, cycle):
    newline = cycle % 40 == 0

    # middle
    sprite_binary_mask = (1 << (39-x))
    # left
    sprite_binary_mask |= (1 << (40-x))
    # right (abs to prevent -1)
    sprite_binary_mask |= (1 << abs(38-x))

    # first cycle is 1, so subtract 1
    x_pos = (cycle-1) % 40

    cycle_binary_mask = (1 << (39-x_pos))

    # copied and pasted these ascii symbols to make it clearer
    if sprite_binary_mask & cycle_binary_mask:
        return "█\n" if newline else "█"
    return " \n" if newline else " "

current_cycle = 1
x = 1
for instruction in data:
    if instruction == "noop":
        print(what_to_print(x, current_cycle), end="")
        current_cycle += 1

    elif instruction.startswith("addx"):
        print(what_to_print(x, current_cycle), end="")
        current_cycle += 1
        print(what_to_print(x, current_cycle), end="")
        current_cycle += 1

        amount = int(instruction[5:])
        x += amount

File: Rust/src/bin/10.rs
// https://adventofcode.com/2022/day/10

use std::fs::File;
use std::io::Read;

fn what_to_print(x: i16, cycle: i16) -> String {
    let newline: bool = cycle % 40 == 0;

    let mut sprite_binary_mask = (1 << (39-x)) as u32;
    sprite_binary_mask |= (1 << (40-x)) as u32;
    sprite_binary_mask |= (1 << (38-x).abs()) as u32;

    let x_pos = (cycle-1) % 40;

    let cycle_binary_mask = (1 << (39-x_pos)) as u32;

    if (sprite_binary_mask & cycle_binary_mask) > 0 {
        return match newline {
            true => "█\n".to_string(),
            false => "█".to_string()
        }
    }
    match newline {
        true => " \n".to_string(),
        false => " ".to_string()
    }
}

fn main() {
    let mut file = File::open("../data/10.txt").expect("Data file not found");
    let mut string_buf = String::new();
    file.read_to_string(&mut string_buf).expect("Reading file failed");
    let lines: Vec<&str> = string_buf.split("\r\n").collect();

    let mut signals: Vec<i16> = vec![];
    // using signed ints make math easier
    let mut current_cycle: i16 = 1;
    let mut x: i16 = 1;

    for &instruction in &lines {
        if instruction == "noop" {
            signals.push(x * current_cycle);
            current_cycle += 1;
        }
        else if instruction.starts_with("addx") {
            signals.push(x * current_cycle);
            current_cycle += 1;
            signals.push(x * current_cycle);
            current_cycle += 1;
            let amount: i16 = (instruction[5..]).parse().unwrap();
            x += amount;
        }
    }

    println!("Part 1");
    println!("{}", signals[19] + signals[59] + signals[99] + signals[139] + signals[179] + signals[219]);

    println!("Part 2");

    let mut current_cycle = 1;
    let mut x = 1;
    for &instruction in &lines {
        if instruction == "noop" {
            print!("{}", what_to_print(x, current_cycle));
            current_cycle += 1;
        }

        else if instruction.starts_with("addx") {
            print!("{}", what_to_print(x, current_cycle));
            current_cycle += 1;
            print!("{}", what_to_print(x, current_cycle));
            current_cycle += 1;

            let amount: i16 = instruction[5..].parse().unwrap();
            x += amount;
        }
    }
}
